tests/lib/rules/no-redeclare.js

```diff
@@ -739,3 +739,592 @@ ruleTester.run("no-redeclare", rule, {
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+				{
+					data: {
+						id: "Object",
+					},
+					messageId: "redeclaredAsBuiltin",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: { ecmaVersion: 6, sourceType: "module" },
+			},
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: {
+					ecmaFeatures: { globalReturn: true },
+					ecmaVersion: 6,
+				},
+			},
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: false }],
+			languageOptions: { parserOptions: { ecmaVersion: 6 } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+
+		// Notifications of readonly are moved from no-undef: https://github.com/eslint/eslint/issues/4504
+		{
+			code: "/*global b:false*/ var b = 1;",
+			options: [{ builtinGlobals: true }],
+			errors: [
+				{
+					data: {
+						id: "b",
+					},
+					messageId: "redeclaredBySyntax",
+				},
+			],
+		},
+
+		{
+			code: `
+	  type T = 1;
+	  type T = 2;
+			`,
+			errors: [
+				{
+					data: {
+						id: "T",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  type NodeListOf = 1;
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: {
+					lib: ["dom"],
+					sourceType: "script",
+				},
+			},
+			errors: [
+				{
+					data: {
+						id: "NodeListOf",
+					},
+					messageId: "redeclaredAsBuiltin",
+				},
+			],
+		},
+		{
+			code: `
+	    var Foo;
+		namespace Foo {
+			export const a = 2;
+		}
+		namespace Foo {
+			export const b = 4;
+		}
+			`,
+			errors: [
+				{
+					data: {
+						id: "Foo",
+					},
+					line: 3,
+					messageId: "redeclared",
+				},
+				{
+					data: {
+						id: "Foo",
+					},
+					line: 6,
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+			namespace Foo {
+				export const a = 1;
+			}
+
+			namespace Bar {
+				export const a = 1;
+			}
+			`,
+			languageOptions: {
+				globals: {
+					Foo: "readonly",
+					Bar: "readonly",
+				},
+			},
+			errors: [
+				{
+					data: {
+						id: "Foo",
+					},
+					line: 2,
+					messageId: "redeclaredAsBuiltin",
+				},
+				{
+					data: {
+						id: "Bar",
+					},
+					line: 6,
+					messageId: "redeclaredAsBuiltin",
+				},
+			],
+		},
+		{
+			code: `
+			/* global Foo */
+			namespace Foo {
+				export const a = 1;
+			}
+
+			/* global Bar */
+			namespace Bar {
+				export const a = 1;
+			}
+			`,
```

Comment:
Same here: this case should be valid, because a `global` comment declares values, whereas `namespace` declares a type.